### DISCUSSION TITLE
How to handle local dates in the client (browser timezone) with UTC dates on the server? (Hydration mismatch / Text content did not match)

### DISCUSSION THREAD

**Discussion by bennettdams** [vercel/next.js#37877] [Help] [answered]
I would like to ask for a case where a mismatch is not unexpected, but rather **expected**: Local dates.

There are some discussions around problems related to hydration mismatches, the typical "Text content did not match" problem. With React 18, this throws a real error instead of being a `console.error`, so a lot more devs will notice it. e.g. https://github.com/vercel/next.js/discussions/35773#discussioncomment-2510947

---

In my app, I want to show the time in the browser timezone. But the static/server pages are created on the server (via `getStaticProps`/`getServerSideProps`) with UTC timzezone. For me, this is an expected mismatch:

![image](https://user-images.githubusercontent.com/29319414/174776666-60cac6e2-c45d-4d8b-8af2-612647cd3272.png)

To my knowledge, there are a few ways to handle this:

- Use Next.js' [Internationalized Routing](https://nextjs.org/docs/advanced-features/i18n-routing) to create different pages. This assumes you're using the language identifier in the routes (like `../en/..`), which then are used for creating the pages with the matching timezone. Using locale identifiers in the routes is not something that everyone wants to do though.
- When rendering a date, render `null` on the server and only create the formatted date on the client (e.g. via `useEffect`). This brings "flickering" with it, as content that was not there on the server now has to be there on the client.
- Set the timezone of the server statically to the one that is expected in the browser. This obviously only works when you're only serving to one specific timezone.

...but all of these are not perfect - I feel like I'm missing something?

Does anyone have a better way or other proposals on how to handle this?

**Comment by icyJoseph** [comment, marked as answer]
- Render the server time, and update with a useEffect to the client time, you could even combine this with CSS visibility set to `hidden` 

```css
.invisible {
 visibility: hidden;
}
```

Elements with hidden visibility still take up space, so no layout shift is triggered. Once you are in the client you can remove the hidden visibility and calculate the correct date.

I think the above is the lowest cost solution. You could also make sure that the server renders UTC time, and it is clear about it, so that people who don't use JS can still calculate it themselves. 

You can just include a `noscript`:

```html
<noscript>
  <style>
    .invisible {
      visibility: visible;
    }
   </style>
</noscript>
```

I'll go over my notes and other projects I've done before, but I think, as said above, that this is the lowest cost solution.

**Reply by icyJoseph** [reply]
I had the chance to expand this with code, on this discussion: https://github.com/vercel/next.js/discussions/38263

**Reply by bennettdams** [reply]
Thanks!

**Comment by douglasrcjames** [comment]
If anyone else runs across this on a web search, if you want the client side time instead of the server side UTC time, doesn't need to have `useEffect` for my use case, just need to run my Javascript function `const dateObject = new Date(timestamp);` within the `"use client";` defined component (.jsx/.tsx) instead of importing it from an external, non-component JavaScript file (.js/.ts).

**Comment by janvorwerk** [comment]
> If anyone else runs across this on a web search, if you want the client side time instead of the server side UTC time, doesn't need to have `useEffect` for my use case, just need to run my Javascript function `const dateObject = new Date(timestamp);` within the `"use client";` defined component (.jsx/.tsx) instead of importing it from an external, non-component JavaScript file (.js/.ts).

It's true, however, you'll be facing the hydratation issue mentionned above, because the client component is still rendered first on the server an then hydrated on the client. Wouldn't you?
https://nextjs.org/docs/messages/react-hydration-error

**Reply by douglasrcjames** [reply]
I see your logic, and I get hydration errors thrown constantly with Next.js in development, but my usage for grabbing the time isn't throwing an error for me with my current setup!
